### PR TITLE
Converts Phil's tranform.py to a lambda convert function for apex

### PR DESCRIPTION
**Pull request by richmahn**
Lambda-fied and Apex-ified this repo. Will be able to convert more than just obs in the future.

---

This change is [Reviewable](https://reviewable.io/reviews/unfoldingword-dev/tx-md2html/5)

### CODE REVIEW

File: functions/convert/main.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            finally:
+                print('finished.')
+
+        except Exception as e:
+            print_error(e.message)
+            self.errors.append(e)
+
+
+def handle(e, ctx):
+    global event, context, data
```
Comment (ethantkoenig) [inline functions/convert/main.py]: Similar to above comment, why is `data` global?

File: functions/convert/main.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import boto3
+import datetime
+
+from glob import glob
+from general_tools.file_utils import unzip, add_file_to_zip, make_dir, write_file
+from general_tools.print_utils import print_error, print_ok
+from general_tools.url_utils import download_file
+
+event = {}
+context = {}
```
Comment (ethantkoenig) [inline functions/convert/main.py]: It's not clear to me what purpose these global variables serve. They aren't used in `run(self)`, and are essentially used as local variables in `handle(..)` (i.e. they are written to before being read from). Further, they aren't used in other modules (at least as far as I can tell). Am I missing something, or can we get rid of these?

**Comment by phillip-hopper** [comment]
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

_[a discussion](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQjicdtfHy38OX9m0_:-KQQjicdtfHy38OX9m0a:bvxdb10) (no related file):_
You will need to add a `.coveragerc` file in order to tell coveralls not to include the requirements directories, otherwise your coverage will be in the single digits.  See https://github.com/unfoldingWord-dev/slack-lambda/blob/develop/.coveragerc

---

_[functions/convert/main.py, line 15 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQi9aIHoNUbiGpOBDl:-KQQi9aIHoNUbiGpOBDm:b-f1mws) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/convert/main.py#L15)):_

> ``` Python
> import markdown
> import boto3
> import datetime
> ```

`datetime` is not used.

---

_[functions/convert/main.py, line 19 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQiE1i1tNDi2OkpoYj:-KQQiE1i1tNDi2OkpoYk:bbr9j6f) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/convert/main.py#L19)):_

> ``` Python
> from glob import glob
> from general_tools.file_utils import unzip, add_file_to_zip, make_dir, write_file
> from general_tools.print_utils import print_error, print_ok
> ```

`print_ok` is not used

---

_[functions/convert/main.py, line 23 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQgGr4NYL4oK1uuT--:-KQQgGr4NYL4oK1uuT-0:b2khgtb) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/convert/main.py#L23)):_

> ``` Python
> 
> event = {}
> context = {}
> ```

`event` and `context` are only used in handle function, and can be removed completely.

---

_[functions/convert/main.py, line 115 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQiSRo3ag5Rcfzk7lY:-KQQiSRo3ag5Rcfzk7lZ:b-7j1sy0) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/convert/main.py#L115)):_

> ``` Python
>     context = ctx
> 
>     if not 'data' in event:
> ```

It is preferred to use `if 'data' not in event:` to test for membership.

---

_[functions/register/main.py, line 10 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQixSgjckQ03rTNN9_:-KQQixShpTk8BlOcxAEz:b2nreh6) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/register/main.py#L10)):_

> ``` Python
> 
> def handle(event, ctx):
>     if not 'tx_manager_url' in event:
> ```

Same as above.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5)_

File: functions/convert/main.py (hunk 1 above)
Comment (richmahn) [reply functions/convert/main.py]: Yeah, probably doesn't need to be global. I wasn't sure if I would need to refer to the payload in the transformOBS class, or other classes to come, as that will eventually have a link to a CSS file, and so forth, but maybe will just stick in in the options variable and not make it global. Also, the context from Lambda holds a unique ID for this Lambda call that can be used, but not necessary. I'll change this to not use global variables. Thanks.

File: functions/convert/main.py (hunk 2 above)
Comment (richmahn) [reply functions/convert/main.py]: @ethantkoenig Great points. I had done this in other lambda functions I ran just so everything has access to the request, both the event variable (which has the data key) and the context variable which holds info on the process. We could get rid of these, yet I always like to have them available if needed in further development. I am sure you have learned different/newer programming theory than us old(er) guys...but do you think it poses a problem? Is it best to just pass everything needed as arguments to a function?

**Comment by phillip-hopper** [comment]
@richmahn I recently pulled the OBS classes into a separate library. It might fix the obs missing error.

File: functions/convert/main.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                    html_template = html_file.read()
+
+                completeHtml = ''
+                for filename in files_to_process:
+                   # read the markdown file
+                    with codecs.open(filename, 'r', 'utf-8-sig') as md_file:
+                        md = md_file.read()
+                        html = markdown.markdown(md)
+                        completeHtml += html
+                        html = TransformOBS.dirRe.sub(r'\1\n' + html + r'\n\2', html_template)
```
Comment (ethantkoenig) [inline functions/convert/main.py]: Is there a reason we need to do this via regular expressions? We are just inserting the `html` variable into a prespecified location of the html template. I would suggest instead having a helper function of the flavor

```
def create_file(html):
    return """<big_template>...<html_here>{}</html_here>...</big_template>""".format(html)
```

I believe this offers a few advantages over the current approach:
1. With the current regular expression approach, we have to make sure that the file `template.html` and `dirRe` are kept in sync. If we decided to change the name of the `obs-content` class in `template.html`, and forget to update `dirRe`, all of a sudden everything will break.
2. I think the suggested approach is easier to understand at a glance than the current approach. (I, for one, had to look at the `re` module to remind myself of the order of the arguments to `sub(..)`)

Comment (phillip-hopper) [reply functions/convert/main.py]: @ethantkoenig I prefer to have the template as a separate file. The major benefit is being able to easily validate the html at design time. This takes advantage of some of the major benefits offered by modern IDE's like IntelliJ.

Comment (ethantkoenig) [reply functions/convert/main.py]: Fair points. Fortunately, the problems I outlined above are mostly due to unnecessarily using regular expressions, not keeping the HTML in a separate file. It's possible to keep the HTML in a separate file while still avoid the problems I outlined above. One option would be,

```
import string
with open("template.html") as template_file:
    template = string.Template(template_file.read())

html = ...
html = template.substitute(content=html)
```

Comment (phillip-hopper) [reply functions/convert/main.py]: @ethantkoenig, I like it. :+1:

**Comment by phillip-hopper** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

_[functions/register/main.py, line 26 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQhaS8CJCWnH8AUHtXV:-KQhaS8CJCWnH8AUHtXW:buhcibw) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/bf2ffa9ca59e63ddaaf098ac74d6181040a74efa/functions/register/main.py#L26)):_

> ``` Python
>     response = requests.post(post_url, data=post_data, headers={'content-type': 'application/json'})
>     return json.loads(response.text)
> ```

Code style standards call for a line feed at the end of python files.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5)_

**Comment by richmahn** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

_[a discussion](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQjicdtfHy38OX9m0_:-KQn_sm_CTItRToWB7KX:bgacjf1) (no related file):_
<details><summary>

<i>Previously, phillip-hopper (Phil Hopper) wrote…</i></summary>



> You will need to add a `.coveragerc` file in order to tell coveralls not to include the requirements directories, otherwise your coverage will be in the single digits.  See https://github.com/unfoldingWord-dev/slack-lambda/blob/develop/.coveragerc
> </details>
> What exactly is coveralls looking at? Every single file? Does requirements.txt also need to be in the omit? What about  directories created with pip install such as functions/convert/uw_tools-0.0.3-py2.7.egg-info/
>  and functions/convert/requests-2.11.1.dist-info/? Do they need to be in the omit section of the .coveragerc file as well?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5)_

**Comment by richmahn** [comment]
@phillip-hopper  You wrote "I recently pulled the OBS classes into a separate library. It might fix the obs missing error."

I have no idea what you mean by this. Not getting any errors when I deploy to apex or send a request for it to convert OBS.

**Comment by phillip-hopper** [comment]
Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

_[a discussion](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQQjicdtfHy38OX9m0_:-KQqlwxwcNXk6bG3SeiC:b4t6fbu) (no related file):_
<details><summary>

<i>Previously, richmahn (Richard Mahn) wrote…</i></summary>



> What exactly is coveralls looking at? Every single file? Does requirements.txt also need to be in the omit? What about  directories created with pip install such as functions/convert/uw_tools-0.0.3-py2.7.egg-info/
>  and functions/convert/requests-2.11.1.dist-info/? Do they need to be in the omit section of the .coveragerc file as well?
> </details>
> Looking at the `.coveragerc` file you added, coveralls will look at every code file in the `functions` directory and all sub directories. It will ignore the `.txt` and `.json` files.

---

_[a discussion](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQqoaFAVd3RhWe6gFPH:-KQqoaFBKXxtKJST5Z6W:b-at6not) (no related file):_
Another note; it looks like there are no unit tests.

---

_[.travis.yml, line 16 [r3]](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5#-KQqnbDwOkWZSWdyg3NS:-KQqnbDx2YYnQagzTCF5:b-dyam38) ([raw file](https://github.com/unfoldingword-dev/tx-md2html/blob/894014838a58dbf3f7fd5c2d59b8bc3f8f0c2526/.travis.yml#L16)):_

> ``` YAML
> - pip install coveralls
> script:
> - coverage run --source=obs test-setup.py test
> ```

You need to remove `--source=obs`. That's what is causing the current failure.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/unfoldingword-dev/tx-md2html/5)_

**Pull request merged**